### ISSUE TITLE
Bug in moment.d.ts with strictNullChecks in tsconfig.json

### ISSUE THREAD

**Issue by SalvatorePreviti** [moment/moment#3599]
**Description of the Issue and Steps to Reproduce:**

With typescript setting "strictNullChecks": true in tsconfig, moment.d.ts fails to compile.

The problem is in the interface CalendarSpec.

```ts
type CalendarSpecVal = string | ((m?: Moment, now?: Moment) => string);
  interface CalendarSpec {
    sameDay?: CalendarSpecVal;
    nextDay?: CalendarSpecVal;
    lastDay?: CalendarSpecVal;
    nextWeek?: CalendarSpecVal;
    lastWeek?: CalendarSpecVal;
    sameElse?: CalendarSpecVal;
    // any additonal properties might be used with moment.calendarFormat
    [x: string]: CalendarSpecVal;
  }
```

Solving the issue is easy, just change the declaration of [x: string] accessor as being possibly undefined in this way:

```ts
[x: string]: CalendarSpecVal | undefined;
```
(line 71)

The fix is very easy and straightforward.

*Setting in tsconfig.json.*

"strictNullChecks": true in tsconfig.json

**Environment:**

TypeScript 2.0.9

Compilation output:
```
63     sameDay?: CalendarSpecVal;
       ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/moment/moment.d.ts(63,5): error TS2411: Property 'sameDay' of type 'string | ((m?: Moment | undefined, now?: Moment | undefined) => string) | undefined' is not assignable to string index type 'CalendarSpecVal'.


64     nextDay?: CalendarSpecVal;
       ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/moment/moment.d.ts(64,5): error TS2411: Property 'nextDay' of type 'string | ((m?: Moment | undefined, now?: Moment | undefined) => string) | undefined' is not assignable to string index type 'CalendarSpecVal'.


65     lastDay?: CalendarSpecVal;
       ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/moment/moment.d.ts(65,5): error TS2411: Property 'lastDay' of type 'string | ((m?: Moment | undefined, now?: Moment | undefined) => string) | undefined' is not assignable to string index type 'CalendarSpecVal'.


66     nextWeek?: CalendarSpecVal;
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/moment/moment.d.ts(66,5): error TS2411: Property 'nextWeek' of type 'string | ((m?: Moment | undefined, now?: Moment | undefined) => string) | undefined' is not assignable to string index type 'CalendarSpecVal'.


67     lastWeek?: CalendarSpecVal;
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/moment/moment.d.ts(67,5): error TS2411: Property 'lastWeek' of type 'string | ((m?: Moment | undefined, now?: Moment | undefined) => string) | undefined' is not assignable to string index type 'CalendarSpecVal'.


68     sameElse?: CalendarSpecVal;
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/moment/moment.d.ts(68,5): error TS2411: Property 'sameElse' of type 'string | ((m?: Moment | undefined, now?: Moment | undefined) => string) | undefined' is not assignable to string index type 'CalendarSpecVal'.
```

**Comment by felixfbecker** [comment]
Guys, this is a serious issue. It was introduced in a non-major version and causes builds to fail everywhere.
https://github.com/felixfbecker/vscode-php-debug/pull/84

**Comment by butterflyhug** [comment]
I think this is a duplicate of #3589 / #3591?

**Comment by duncfro** [comment]
Yep, this broke our builds too (at the same time as a nasty nodemailer/aws-sdk type definition problem... things aren't very stable in the world of TypeScript yet!). Staying at moment version 2.15.2 is fine for us.

**Comment by itaysabato** [comment]
Please fix

**Comment by itaysabato** [comment]
Seems like #3591 fixed this issue - installed version `2.17.0` which seems to work properly.

**Comment by maggiepint** [comment]
closing
